mitodl/ocw-studio
# Resource links without forking CKEditor's Link Plugin

#### Pre-Flight checklist


#### What are the relevant tickets?
#1619 

#### What's this PR do?
This PR changes how we implement CKEditor support for resource links. Resource links are very similar to regular links in CKEditor. The main differences are markdown conversion and the fact that the default text for a regular link is its `href`, but the default text for a resource link is the resource title.

- **Previously**: We [forked CKEditor](https://github.com/mitodl/ckeditor5) and edited its link plugin to accommodate the differences. Resource link uuid was stored as `data-uuid` attribute on the anchor.
- **Now:** We just use the regular link plugin and store the uuid as a query parameter value. The custom behavior we want is achieved by new plugins rather than editing a forked plugin.

**Two behavior changes:**
- Previously, if you clicked on a resource link it would show "no href". Now it shows an href that links to that specific resource's page in studio. *Using the actual link plugin, we need to show some href, so this was a natural choice*.
- Previously, you could edit (click link, click ✏️ button) the href of a resource link which...broke stuff. That was really a bug. It's fixed now. 

#### How should this be manually tested?
Try things like:
- inserting a new resource link
    - with no text highlighted... link title should be resource title
    - with text highlighted ... text should be converted to a resource link
- editing the text of a resource link
    - adding characters to the end of a resource link either as part of the link or as part of the following non-link text. 
    - adding/removing characters mid-link
- subscripts/superscripts in a resource link
- removing a resource link by clicking it and hitting the "X" button
- try to edit the resource link href by clicking the link and hitting ✏️. (It should be disabled)

#### Where should the reviewer start?
Start with docstrings on `ResourceLink`, `ResourceLinkMarkdownSyntax`, and `ResourceLinkUI` plugins.

#### Screenshots (if appropriate)
<img width="1107" alt="Screenshot 2023-01-12 at 10 54 21 AM" src="https://user-images.githubusercontent.com/9010790/212115896-d79b9d35-46ed-491d-937b-9fbe739ccd31.png">
